File: ios/Ember/Core/Utils/RetryHelper.swift
import Foundation

/// Retries an async throwing operation with exponential backoff.
///
/// Usage:
/// ```swift
/// let result = try await RetryHelper.withExponentialBackoff {
///     try await apiClient.request(endpoint: .listCharacters, responseType: CharacterListResponse.self)
/// }
/// ```
enum RetryHelper {
    /// Default configuration for retry behavior.
    struct Configuration {
        /// Maximum number of retry attempts (not counting the initial attempt).
        let maxRetries: Int

        /// Base delay between retries (doubles each attempt).
        let baseDelay: TimeInterval

        /// Maximum delay cap.
        let maxDelay: TimeInterval

        /// Whether to add jitter (random variation) to delays.
        let jitter: Bool

        static let `default` = Configuration(
            maxRetries: 3,
            baseDelay: 1.0,
            maxDelay: 16.0,
            jitter: true
        )

        static let aggressive = Configuration(
            maxRetries: 5,
            baseDelay: 0.5,
            maxDelay: 30.0,
            jitter: true
        )

        static let gentle = Configuration(
            maxRetries: 2,
            baseDelay: 2.0,
            maxDelay: 8.0,
            jitter: false
        )
    }

    /// Executes the operation, retrying on failure with exponential backoff.
    ///
    /// Only retries for errors where `EmberError.isRetryable` is `true`.
    /// Non-retryable errors (unauthorized, not found, decoding) throw immediately.
    ///
    /// - Parameters:
    ///   - configuration: Retry timing configuration.
    ///   - operation: The async throwing closure to execute.
    /// - Returns: The result of a successful operation.
    /// - Throws: The last error if all retries are exhausted, or immediately for non-retryable errors.
    static func withExponentialBackoff<T>(
        configuration: Configuration = .default,
        operation: () async throws -> T
    ) async throws -> T {
        var lastError: Error?
        var attempt = 0

        while attempt <= configuration.maxRetries {
            do {
                return try await operation()
            } catch {
                lastError = error

                let emberError = EmberError.from(error)

                // Non-retryable errors throw immediately
                guard emberError.isRetryable else {
                    throw error
                }

                // Exhausted retries
                if attempt == configuration.maxRetries {
                    break
                }

                // Calculate delay with exponential backoff
                let delay = calculateDelay(
                    attempt: attempt,
                    baseDelay: configuration.baseDelay,
                    maxDelay: configuration.maxDelay,
                    jitter: configuration.jitter
                )

                try await Task.sleep(nanoseconds: UInt64(delay * 1_000_000_000))
                attempt += 1
            }
        }

        throw lastError ?? EmberError.unknown("Retry failed")
    }

    // MARK: - Private

    private static func calculateDelay(
        attempt: Int,
        baseDelay: TimeInterval,
        maxDelay: TimeInterval,
        jitter: Bool
    ) -> TimeInterval {
        // Exponential: baseDelay * 2^attempt
        let exponentialDelay = baseDelay * pow(2.0, Double(attempt))
        let cappedDelay = min(exponentialDelay, maxDelay)

        if jitter {
            // Add random jitter: 50% to 100% of the calculated delay
            let jitterFraction = Double.random(in: 0.5...1.0)
            return cappedDelay * jitterFraction
        }

        return cappedDelay
    }
}

File: android/app/src/main/java/ai/ember/app/core/error/RetryHelper.kt
package ai.ember.app.core.error

import kotlinx.coroutines.delay
import kotlin.math.min
import kotlin.random.Random

/**
 * Exponential backoff retry utility for transient errors.
 *
 * Only retries operations that fail with [EmberError.isRetryable] errors.
 * Non-retryable errors are thrown immediately without consuming attempts.
 *
 * Mirrors iOS RetryHelper behavior.
 */
object RetryHelper {

    /**
     * Executes [action] with exponential backoff retry logic.
     *
     * @param maxAttempts Maximum number of attempts (including initial).
     * @param baseDelayMs Base delay in milliseconds for exponential backoff.
     * @param maxDelayMs Maximum delay cap in milliseconds.
     * @param jitter Whether to add random jitter to delay.
     * @param action The suspending block to execute and potentially retry.
     * @return The result of [action] on success.
     * @throws EmberErrorException if all retries are exhausted or a non-retryable error occurs.
     */
    suspend fun <T> withRetry(
        maxAttempts: Int = DEFAULT_MAX_ATTEMPTS,
        baseDelayMs: Long = DEFAULT_BASE_DELAY_MS,
        maxDelayMs: Long = DEFAULT_MAX_DELAY_MS,
        jitter: Boolean = true,
        action: suspend () -> T,
    ): T {
        var lastError: EmberError? = null

        repeat(maxAttempts) { attempt ->
            try {
                return action()
            } catch (e: Exception) {
                val emberError = when (e) {
                    is EmberErrorException -> e.emberError
                    else -> EmberError.from(e)
                }

                if (!emberError.isRetryable) {
                    throw EmberErrorException(emberError)
                }

                lastError = emberError

                if (attempt < maxAttempts - 1) {
                    val delayMs = calculateDelay(attempt, baseDelayMs, maxDelayMs, jitter)
                    delay(delayMs)
                }
            }
        }

        throw EmberErrorException(lastError ?: EmberError.Unknown())
    }

    /**
     * Calculates the backoff delay for a given attempt.
     *
     * Uses exponential backoff: base * 2^attempt, capped at maxDelay.
     * Jitter adds random variance between 0 and the calculated delay.
     */
    internal fun calculateDelay(
        attempt: Int,
        baseDelayMs: Long,
        maxDelayMs: Long,
        jitter: Boolean,
    ): Long {
        val exponentialDelay = baseDelayMs * (1L shl attempt)
        val cappedDelay = min(exponentialDelay, maxDelayMs)
        return if (jitter) {
            Random.nextLong(cappedDelay / 2, cappedDelay + 1)
        } else {
            cappedDelay
        }
    }

    private const val DEFAULT_MAX_ATTEMPTS = 3
    private const val DEFAULT_BASE_DELAY_MS = 1000L
    private const val DEFAULT_MAX_DELAY_MS = 10000L
}
